### PR TITLE
Search for person by NOMIS ID for CAS2

### PR DESCRIPTION
**Pull request by Harriethw**
This depends on the AP Tools PR being merged: https://github.com/ministryofjustice/hmpps-approved-premises-tools/pull/51 
If you'd like to call the added controller function from the UI there is this WIP PR: https://github.com/ministryofjustice/hmpps-community-accommodation-tier-2-ui/pull/310

We need to allow a CAS2 user to search for an offender by prison number (nomsNumber), from an API that returns a CRN in order to make subsequent requests for OASyS data. This PR makes a request to [the `probation-offender-search` API](https://github.com/ministryofjustice/probation-offender-search). The shape of the [data returned should match what's returned from the probation API](http://localhost:8084/swagger-ui/index.html#/offender-search/search_1), which is a list of OffenderDetails.

See their controller tests:
https://github.com/ministryofjustice/probation-offender-search/blob/7093fdf716f969de3b80827675f8389cef01f95c/src/test/kotlin/uk/gov/justice/hmpps/probationsearch/controllers/OffenderSearchControllerTest.kt
And Swagger: http://localhost:8084/swagger-ui/index.html#/offender-search/search_1 

I have some outstanding questions on the implementation:

~~1. If an offender is an LAO, we don't want to allow the user to continue. If I understand correctly if `currentExclusion` or `currentRestriction` are `true` then the offender is an LAO? In which case we can just return a Forbidden response? I'm not entirely sure what those fields mean on this probation offender search API.~~ the implementation for LAO should mean a forbidden response is returned if they are LAO
2. The current `FullPerson` requires a date of birth, which according to the Probation API model might potentially be `null`. How should we handle that? Do we want to continue on CAS2 if there is no DoB, as it could potentially be the wrong person? Or, should there be the concept of a CAS2 Person that might not have a DoB? Orrrr, do we make another call to another Delius API to get that DoB.
~~3. We don't get an `inOutStatus` from the Probation API. Can we assume the person is in custody, because the POM is making an application for CAS2? Would this have implications if it was wrong?~~ we are making call to Prison API to get extra prison info
4. Are we making the right call to the `probation-offender-search` API, and have I understood the shape of the data we should expect back?

### CODE REVIEW

File: src/main/resources/static/codegen/built-cas2-api-spec.yml [hunk 1]
```diff
@@ -241,11 +241,11 @@ paths:
     get:
       tags:
         - People operations
-      summary: Searches for a Person by their CRN
+      summary: Searches for a Person by their Prison Number (NOMIS ID)
```
Comment (edavey) [inline src/main/resources/static/codegen/built-cas2-api-spec.yml:244]: Would it be clearer if this were part of the same commit where `cas2-api.yml` is modified?

**Review by edavey — COMMENTED** [review]
this looks great, thks! Just some little comments and suggestions. It does look like we need to make the call to the prison side to find the name of the prison where the person is located.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/client/ProbationOffenderSearchApiClient.kt [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+@Component
+class ProbationOffenderSearchApiClient(
+  @Qualifier("probationOffenderSearchApiWebClient") webClient: WebClient,
+  objectMapper: ObjectMapper,
+  webClientCache: WebClientCache,
+) : BaseHMPPSClient(webClient, objectMapper, webClientCache) {
+
+  fun searchOffenderByNomsNumber(nomsNumber: String) = postRequest<List<ProbationOffenderDetail>> {
+    path = "/search"
```
[the remark this answers is not in the document]
Comment (Harriethw) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/client/ProbationOffenderSearchApiClient.kt:18]: hmm, maybe it's one to chat about as I've just seen there is also a `nomsNumber` endpoint! https://probation-offender-search-dev.hmpps.service.justice.gov.uk/swagger-ui/index.html#/offender-search/findByNomsNumbers 

I think I'd prefer it to be more explicit, especially having worked with opensearch APIs before I've had trouble getting exact matches (that might have been an issue with that API though!)

File: src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/unit/service/cas2/OffenderServiceTest.kt [hunk 3]
```diff
@@ -44,9 +49,10 @@ class OffenderServiceTest {
   }
 
   private val offenderService = OffenderService(
-    mockCommunityApiClient,
-    mockPrisonsApiClient,
-    mockApOASysContextApiClient,
+          mockCommunityApiClient,
+          mockPrisonsApiClient,
+          mockProbationOffenderSearchClient,
+          mockApOASysContextApiClient,
```
[the remark this answers is not in the document]
Comment (Harriethw) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/unit/service/cas2/OffenderServiceTest.kt:55]: no, I think something has changed about my Kotlin linting in the IDE...

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/transformer/PersonTransformer.kt [hunk 4]
```diff
@@ -68,6 +69,21 @@ class PersonTransformer {
...
+  fun transformProbationOffenderToPersonApi(probationOffenderResult: ProbationOffenderSearchResult.Success.Full, nomsNumber: String) =
+          FullPerson(
+                  type = PersonType.fullPerson,
+                  crn = probationOffenderResult.probationOffenderDetail.otherIds.crn,
+                  name = "${probationOffenderResult.probationOffenderDetail.firstName} ${probationOffenderResult.probationOffenderDetail.surname}",
+                  // TODO: we are not handling if DoB is not there, what should we do?
+                  dateOfBirth = probationOffenderResult.probationOffenderDetail.dateOfBirth!!,
+                  sex = probationOffenderResult.probationOffenderDetail.gender ?: "Not Found",
+                  // TODO: do we need to make call to prison api to determine status?
+                  status = FullPerson.Status.inCustody,
```
[the remark this answers is not in the document]
Comment (Harriethw) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/transformer/PersonTransformer.kt]: brill, thanks - have added that call now.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/transformer/PersonTransformer.kt [hunk 5]
```diff
@@ -68,6 +69,21 @@ class PersonTransformer {
     )
   }
 
+  fun transformProbationOffenderToPersonApi(probationOffenderResult: ProbationOffenderSearchResult.Success.Full, nomsNumber: String) =
+          FullPerson(
+                  type = PersonType.fullPerson,
+                  crn = probationOffenderResult.probationOffenderDetail.otherIds.crn,
+                  name = "${probationOffenderResult.probationOffenderDetail.firstName} ${probationOffenderResult.probationOffenderDetail.surname}",
+                  // TODO: we are not handling if DoB is not there, what should we do?
+                  dateOfBirth = probationOffenderResult.probationOffenderDetail.dateOfBirth!!,
```
[the remark this answers is not in the document]
Comment (Harriethw) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/transformer/PersonTransformer.kt]: hmm `dateOfBirth` is a required field on `FullPerson`... I'm wondering if we let it blow up for now then, or change the Person class? who would we ask to check who's missing dob?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/cas2/OffenderService.kt [hunk 6]
```diff
@@ -7,24 +7,71 @@ import uk.gov.justice.digital.hmpps.approvedpremisesapi.client.ApOASysContextApi
...
 ) {
 
   private val log = LoggerFactory.getLogger(this::class.java)
 
+  fun getPersonByNomsNumber(nomsNumber: String): ProbationOffenderSearchResult {
+    fun logFailedResponse(probationResponse: ClientResult.Failure<List<ProbationOffenderDetail>>) =
+            log.warn("Could not get inmate details for $nomsNumber", probationResponse.toException())
+
+    // TODO: do we need to use cache, as below?
+    val probationResponse = probationOffenderSearchApiClient.searchOffenderByNomsNumber(nomsNumber)
```
Comment (Harriethw) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/cas2/OffenderService.kt:37]: @edavey do you have any thoughts on using the cache?

Comment (edavey) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/cas2/OffenderService.kt:37]: no I don't think we need to cache calls to Probation Offender Search -- as it is in itself a cache of sorts, on top of Delius/Community API...

Comment (pezholio) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/cas2/OffenderService.kt:37]: I think we're leaning towards moving away from so much caching, often the response is faster than Redis anyway!

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/transformer/PersonTransformer.kt [hunk 7]
```diff
@@ -68,6 +69,24 @@ class PersonTransformer {
     )
   }
 
+  fun transformProbationOffenderToPersonApi(probationOffenderResult: ProbationOffenderSearchResult.Success.Full, nomsNumber: String) =
+          FullPerson(
+            type = PersonType.fullPerson,
+            crn = probationOffenderResult.probationOffenderDetail.otherIds.crn,
+            name = "${probationOffenderResult.probationOffenderDetail.firstName} ${probationOffenderResult.probationOffenderDetail.surname}",
+            // TODO: we are not handling if DoB is not there, what should we do?
+            dateOfBirth = probationOffenderResult.probationOffenderDetail.dateOfBirth!!,
```
Comment (pezholio) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/transformer/PersonTransformer.kt:78]: Having that `!!` there makes me nervous. Could you extend the CAS2 `FullPerson` to make `dateOfBirth` optional?

**Review by pezholio — APPROVED** [review]
LGTM from where I'm standing. Just a couple of things to think about

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/transformer/PersonTransformer.kt (hunk 7 above)
Comment (Harriethw) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/transformer/PersonTransformer.kt]: I'm using the shared `FullPerson` model at the moment, I'm wondering if I should create a CAS2 version or if there's a way to extend the `FullPerson` with just the change to that one field

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/client/ProbationOffenderSearchApiClient.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (Harriethw) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/client/ProbationOffenderSearchApiClient.kt:18]: we decided will use the current implementation 👍

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/transformer/PersonTransformer.kt (hunk 7 above)
Comment (Harriethw) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/transformer/PersonTransformer.kt]: after chatting with @edavey we decided to take the risk of having the !! for now as it's very unlikely we won't have DoB (we're assuming!), and we don't want to create yet another model for this one difference to `FullPerson`,  and I'm also thinking it should be required as a piece of data that identifies the right person.

**Pull request merged**